### PR TITLE
Added database writing for fuel quotes

### PR DESCRIPTION
**Pull request by tnttruong**
### CHANGES
- added `fuelquote()` table for handling of fuelquote information
- `store_quote()` function in `database.rs` for database handling of quote form information
- unit test `test_store_quote()` check quote storage

### ISSUES
- no read function for `fuelquote()` to be outputted to `routes.rs`

### CODE REVIEW

File: src/backend/src/config.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ fn default_database_uri() -> String {
 }
 
 fn default_max_connections() -> u32 {
-    1
+    2
```
Comment (Brod8362) [inline src/backend/src/config.rs:11]: This should stay 1.

File: src/backend/src/database.rs [hunk 2]
```diff
@@ -37,6 +43,15 @@ pub async fn init_tables(pool: &Pool<Sqlite>) -> Result<(), sqlx::Error> {
...
     ").await?;
+
+    tx.execute("
+        CREATE TABLE IF NOT EXISTS fuel(
+            gallons_requested INTEGER NOT NULL,
+            address TEXT NOT NULL,
+            delivery_date TEXT NOT NULL
+        );
+    ").await?;
+
```
Comment (Brod8362) [inline src/backend/src/database.rs:54]: - "fuel" is a bad name for this table, it should be something like "quotes" or "fuel_quotes"
- Missing foreign key for user id (who made the request)
- Missing fields: refer to Assignment 2

File: src/backend/src/database.rs [hunk 3]
```diff
@@ -128,6 +143,30 @@ pub async fn authenticate_user(pool: &Pool<Sqlite>, username: &str, password: &s
...
 }
 
+pub async fn store_quote(pool: &Pool<Sqlite>, gallons_requested: &str, address: &str, delivery_date: &str) -> Result<(), Box<dyn Error>>{
+    sqlx::query("SELECT * FROM fuel")
+        .bind(gallons_requested)
+        .bind(address)
+        .bind(delivery_date)
+        .fetch_all(pool)
+        .await?;
+
```
Comment (Brod8362) [inline src/backend/src/database.rs:153]: Why is this here? This is a submission function. The result isn't even used (the compiler should be giving you a warning about this)

File: src/backend/src/database.rs [hunk 4]
```diff
@@ -128,6 +143,30 @@ pub async fn authenticate_user(pool: &Pool<Sqlite>, username: &str, password: &s
 
 }
 
+pub async fn store_quote(pool: &Pool<Sqlite>, gallons_requested: &str, address: &str, delivery_date: &str) -> Result<(), Box<dyn Error>>{
```
Comment (Brod8362) [inline src/backend/src/database.rs:146]: `gallons_requested` should be a float/double.
Make sure datatypes match what they should be, in general

File: src/backend/src/database.rs [hunk 5]
```diff
@@ -137,7 +176,7 @@ mod tests {
 
     #[tokio::test]
     pub async fn test_create_user() {
-        let pool = crate::database::init_connection("sqlite://:memory:", 1).await.unwrap();
+        let pool = crate::database::init_connection("sqlite://:memory:", 2).await.unwrap();
```
Comment (Brod8362) [inline src/backend/src/database.rs:179]: Do not change unit tests not related to your own. If you are doing that, it is because you broke something.

File: src/backend/src/database.rs [hunk 6]
```diff
@@ -153,7 +192,7 @@ mod tests {
 
     #[tokio::test]
     pub async fn test_authenticate_user() {
-        let pool = crate::database::init_connection("sqlite://:memory:", 1).await.unwrap();
+        let pool = crate::database::init_connection("sqlite://:memory:", 2).await.unwrap();
```
Comment (Brod8362) [inline src/backend/src/database.rs:195]: Do not change unit tests not related to your own. If you are doing that, it is because you broke something.

File: src/backend/src/database.rs [hunk 7]
```diff
@@ -167,4 +206,18 @@ mod tests {
         let auth_bad_username = crate::database::authenticate_user(&pool, "wrongusername", "password").await;
         assert!(auth_bad_username.is_err());
     }
+
+    #[tokio::test]
+    pub async fn test_store_quote(){
+        let pool = crate::database::init_connection("sqlite://:memory:", 2).await.unwrap();
```
Comment (Brod8362) [inline src/backend/src/database.rs:212]: Why 2 connections?

File: src/backend/src/routes.rs [hunk 8]
```diff
@@ -176,15 +176,21 @@ pub struct QuoteData {
...
+    let submit_resp = database::store_quote(pool, &form.gallons_requested, &form.address, &form.delivery_date).await;
+    if submit_resp.is_err(){
+        let error_message = format!("{:?}", submit_resp.err().unwrap());
+        return Template::render(
+            "fuel_quote",
+            context!{
+                error: error_message
+            }
+        )
+    }
```
Comment (Brod8362) [inline src/backend/src/routes.rs:188]: Use pattern matching or `if let`, not `if result.is_err()`.  https://doc.rust-lang.org/book/ch18-00-patterns.html

**Review by Brod8362 — CHANGES_REQUESTED** [review]

File: src/backend/src/database.rs (hunk 7 above)
Comment (tnttruong) [reply src/backend/src/database.rs:212]: when tested with only 1 connection it results in a "database is locked" error, which resolves itself when 2 connections are initiated

File: src/backend/src/database.rs [hunk 9]
```diff
@@ -45,10 +39,17 @@ pub async fn init_tables(pool: &Pool<Sqlite>) -> Result<(), sqlx::Error> {
...
-            gallons_requested INTEGER NOT NULL,
+        CREATE TABLE IF NOT EXISTS fuelquote(
+            quote_id TEXT PRIMARY KEY NOT NULL,
+            user_id TEXT NOT NULL,
+            gallons_requested REAL NOT NULL,
             address TEXT NOT NULL,
-            delivery_date TEXT NOT NULL
+            delivery_date TEXT NOT NULL,
+            suggest_price INTEGER NOT NULL,
+            total_due INTEGER NOT NULL,
```
Comment (Brod8362) [inline src/backend/src/database.rs:49]: These two values should be REAL instead of INTEGER

File: src/backend/src/database.rs [hunk 10]
```diff
@@ -209,15 +208,14 @@ mod tests {
 
     #[tokio::test]
     pub async fn test_store_quote(){
-        let pool = crate::database::init_connection("sqlite://:memory:", 2).await.unwrap();
-        let success_store = crate::database::store_quote(&pool, "10", "address1", "2024-03-26").await;
+        let pool = crate::database::init_connection("sqlite://:memory:", 1).await.unwrap();
+        //make a valid user for the test
+        let user_id = crate::database::register_user(&pool, "test", "password").await.unwrap();
+        let _ = crate::database::authenticate_user(&pool, "test", "password").await;
```
Comment (Brod8362) [inline src/backend/src/database.rs:214]: No need to authenticate

File: src/backend/src/routes.rs [hunk 11]
```diff
@@ -169,33 +169,33 @@ pub fn quote() -> Template {
...
-        let error_message = format!("{:?}", submit_resp.err().unwrap());
-        return Template::render(
+pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: Form<QuoteData>) -> Template {
+    if let Some(user_id) = &ctx.0 {
+        let submit_resp = database::store_quote(pool, &user_id, form.gallons_requested, &form.address, &form.delivery_date).await;
+        Template::render(
             "fuel_quote",
             context!{
-                error: error_message
+                //TODO: delete and replace with confirm msg once DB write is done
```
Comment (Brod8362) [inline src/backend/src/routes.rs:184]: Resolve this TODO now that DB write is done

**Review by Brod8362 — CHANGES_REQUESTED** [review]
A couple small changes but overall looks good - make those and it's ready to merge.

File: src/backend/src/routes.rs [hunk 12]
```diff
@@ -181,7 +181,7 @@ pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: F
         Template::render(
             "fuel_quote",
             context!{
-                //TODO: delete and replace with confirm msg once DB write is done
+                //TODO: delete and replace with confirm msg once DB read is done
                 gallons_requested: &form.gallons_requested,
                 address: &form.address,
                 delivery_date: &form.delivery_date
```
Comment (Brod8362) [inline src/backend/src/routes.rs:187]: There's no reason this can't be done now. Just a message that says something along the lines of "quote successful" is good enough.
Also, make sure to actually use `submit_response` - check that things are okay before telling the user they're okay.

**Review by Brod8362 — CHANGES_REQUESTED** [review]
See comments

File: src/backend/src/routes.rs (hunk 12 above)
Comment (Brod8362) [reply src/backend/src/routes.rs:187]: The compiler is most likely giving you a warning about unused variables /unused result - we need to make sure we handle every result. (The only exception is in unit tests)

**Review by Brod8362 — APPROVED** [review]
Looks good, awaiting one other person to review

File: src/backend/src/database.rs [hunk 13]
```diff
@@ -27,9 +27,7 @@ pub enum AuthenticationError {
 #[derive(Error, Debug)]
 pub enum QuoteFormError {
     #[error("No gallons requested")]
-    EmptyGallons,
-    #[error("No date chosen")]
```
Comment (Mnguyener) [inline src/backend/src/database.rs]: What's the reason for removing the error check for `EmptyDate`?

File: src/backend/src/database.rs [hunk 14]
```diff
@@ -143,26 +149,17 @@ pub async fn authenticate_user(pool: &Pool<Sqlite>, username: &str, password: &s
...
-pub async fn store_quote(pool: &Pool<Sqlite>, gallons_requested: &str, address: &str, delivery_date: &str) -> Result<(), Box<dyn Error>>{
-    sqlx::query("SELECT * FROM fuel")
-        .bind(gallons_requested)
-        .bind(address)
-        .bind(delivery_date)
-        .fetch_all(pool)
-        .await?;
-
-    //checks if entered empty amt of gallons
-    if gallons_requested.is_empty(){
```
Comment (Mnguyener) [inline src/backend/src/database.rs]: What's the reason for removing the check for if the `gallons_requested` is empty?

**Review by Mnguyener — COMMENTED** [review]
See my questions

File: src/backend/src/routes.rs [hunk 15]
```diff
@@ -169,27 +169,43 @@ pub fn quote() -> Template {
...
-    //renders form results FOR NOW until database implement
-    //let form_inp = form.into_inner();
-
-    Template:: render(
-        "fuel_quote",
-        context!{
-            gallons_requested: &form.gallons_requested,
-            address: &form.address,
-            delivery_date: &form.delivery_date
+pub async fn quote_submit(ctx: &UserContext, pool: &State<Pool<Sqlite>>, form: Form<QuoteData>) -> Template {
```
Comment (Mnguyener) [inline src/backend/src/routes.rs:178]: I get a "Submission error" message when I try to submit a fuel quote and none of the changes got reflect on the form
image

**Review by Mnguyener — CHANGES_REQUESTED** [review]
Please account for the error in your unit test

**Comment by Brod8362** [comment]
image
It works fine for me.
Can you describe specific steps to reproduce the issue?

File: src/backend/src/database.rs (hunk 13 above)
Comment (Brod8362) [reply src/backend/src/database.rs]: It's not possible for the date to be empty, because if it was not specified then the deserializer would return a 422.

File: src/backend/src/database.rs (hunk 14 above)
Comment (Brod8362) [reply src/backend/src/database.rs]: `gallons_requested` was changed to an `f64` which cannot be empty

**Comment by Mnguyener** [comment]
> image It works fine for me. Can you describe specific steps to reproduce the issue?

So I check out to the branch and pulled all the recent changes. Then, I ran `sqlite3 database.db` and `cargo run`. I check the tables and they exist: `auth` and `fuelquote`. Next, I click on the Fuel Quote History page and type in `20` for gallons and today's date. That is how I got my screenshot. The status responses look fine though
image

**Comment by Brod8362** [comment]
`sqlite3 database.db` will not create the database. Make an empty file with `touch database.db` instead.

**Comment by Brod8362** [comment]
Following those exact steps and I can reproduce the error - I shall investigate

**Comment by Mnguyener** [comment]
> `sqlite3 database.db` will not create the database. Make an empty file with `touch database.db` instead.

Ok, I `rm database.db`, closed my sqlite3 connection, and `touch database.db` then repeated my steps on the website. Yet I still face the same error

**Comment by Brod8362** [comment]
` SqliteError { code: 787, message: "FOREIGN KEY constraint failed", }`

It occurs when the user is not logged in.
This is happening because in general, we aren't checking authentication *anywhere*.

I already have local changes that address this - so I will deal with it in that PR.

**Comment by Mnguyener** [comment]
Despite the fact that it looks like I am logged in (my username shows on the page), I am not actually logged in?

**Comment by Brod8362** [comment]
Probably browser cache issue - force-refresh.

**Comment by Mnguyener** [comment]
I cleared my cookies and registered my account again. I also deleted the previous `db` just in case. Yet I still face the same submission error.

**Comment by Brod8362** [comment]
I see the problem.
`UserContext` actually carriers the user's *name*, not their user ID. But the foreign key is the user_id - so it fails.

I already have authentication changes in my local that address this, waiting on this PR to merge them. So let's skip this particular issue for now and I will fix it in my branch.

**Comment by Mnguyener** [comment]
Awesome, glad we know the problem now

**Review by Mnguyener — APPROVED** [review]
